File: leetcode/0207-course-schedule.js
var canFinish = function(numCourses, prerequisites) {
    // 인접 리스트 생성
    // graph[i] = 과목 i를 선수과목으로 필요로 하는 과목들
    const graph = Array.from({ length: numCourses }, () => []);

    // 각 과목의 진입 차수
    const inDegree = new Array(numCourses).fill(0);

    // 그래프 생성 및 진입 차수 계산
    for (const [course, prerequisite] of prerequisites) {
        // prerequisite → course
        graph[prerequisite].push(course);

        // course의 선수과목 개수 증가
        inDegree[course]++;
    }

    // 선수과목이 없는 과목부터 큐에 추가
    const queue = [];

    for (let courseId = 0; courseId < numCourses; courseId++) {
        if (inDegree[courseId] === 0) {
            queue.push(courseId);
        }
    }

    // 위상 정렬
    for (const currentCourse of queue) {
        // 현재 과목을 처리했으므로 남은 과목 수 감소
        numCourses--;

        // 현재 과목을 선수과목으로 갖는 과목들 처리
        for (const dependentCourse of graph[currentCourse]) {
            inDegree[dependentCourse]--;

            // 모든 선수과목을 이수했다면 큐에 추가
            if (inDegree[dependentCourse] === 0) {
                queue.push(dependentCourse);
            }
        }
    }

    // 모든 과목을 처리했다면 사이클이 없음
    return numCourses === 0;
};

File: leetcode/0207-course-schedule/0207-course-schedule.java
class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        // Create adjacency list to represent the course dependency graph
        // graph[i] contains all courses that depend on course i
        List<Integer>[] graph = new List[numCourses];
        Arrays.setAll(graph, index -> new ArrayList<>());
      
        // Track the in-degree (number of prerequisites) for each course
        int[] inDegree = new int[numCourses];
      
        // Build the graph and calculate in-degrees
        // For prerequisite [a, b]: course a depends on course b
        // So we add edge from b -> a
        for (int[] prerequisite : prerequisites) {
            int course = prerequisite[0];
            int prerequisiteCourse = prerequisite[1];
            graph[prerequisiteCourse].add(course);
            inDegree[course]++;
        }
      
        // Queue to store courses with no prerequisites (in-degree = 0)
        Deque<Integer> queue = new ArrayDeque<>();
      
        // Add all courses with no prerequisites to the queue
        for (int course = 0; course < numCourses; course++) {
            if (inDegree[course] == 0) {
                queue.offer(course);
            }
        }
      
        // Process courses using topological sort (Kahn's algorithm)
        while (!queue.isEmpty()) {
            // Take a course with no remaining prerequisites
            int currentCourse = queue.poll();
            // Decrement the count of remaining courses to process
            numCourses--;
          
            // For each course that depends on the current course
            for (int dependentCourse : graph[currentCourse]) {
                // Reduce its in-degree since we've completed a prerequisite
                inDegree[dependentCourse]--;
                // If all prerequisites are met, add to queue
                if (inDegree[dependentCourse] == 0) {
                    queue.offer(dependentCourse);
                }
            }
        }
      
        // If all courses are processed (numCourses = 0), no cycle exists
        return numCourses == 0;
    }
}
